provider/src/main/java/com/cloudera/director/google/sql/GoogleCloudSQLProvider.java

```diff
@@ -0,0 +1,608 @@
+          throw new TransientProviderException(e);
+      }
+    } catch (IOException e) {
+      throw new TransientProviderException(e);
+    }
+
+    this.resourceTemplateConfigurationValidator =
+        new CompositeConfigurationValidator(METADATA.getResourceTemplateConfigurationValidator(),
+            new GoogleCloudSQLInstanceTemplateConfigurationValidator(this));
+  }
+
+  @Override
+  public ResourceProviderMetadata getProviderMetadata() {
+    return METADATA;
+  }
+
+  @Override
+  public ConfigurationValidator getResourceTemplateConfigurationValidator() {
+    return resourceTemplateConfigurationValidator;
+  }
+
+  @Override
+  public Resource.Type getResourceType() {
+    return AbstractDatabaseServerInstance.TYPE;
+  }
+
+  @Override
+  public GoogleCloudSQLInstanceTemplate createResourceTemplate(
+      String name, Configured configuration, Map<String, String> tags) {
+    return new GoogleCloudSQLInstanceTemplate(name, configuration, tags, getLocalizationContext());
+  }
+
+  @Override
+  public void allocate(GoogleCloudSQLInstanceTemplate template,
+      Collection<String> instanceIds, int minCount) throws InterruptedException {
+
+    PluginExceptionConditionAccumulator accumulator = new PluginExceptionConditionAccumulator();
+
+    LocalizationContext providerLocalizationContext = getLocalizationContext();
+    LocalizationContext templateLocalizationContext =
+        SimpleResourceTemplate.getTemplateLocalizationContext(providerLocalizationContext);
+
+    for (String instanceId : instanceIds) {
+      String decoratedInstanceName = decorateInstanceName(template, instanceId, templateLocalizationContext);
+      if (!instanceNamePrefixPattern.matcher(decoratedInstanceName).matches()) {
+        throw new InterruptedException("The database instance name " + decoratedInstanceName + " doesn't meet Google Cloud SQL requirements. It can only contain lowercase letters, numbers and hyphens and start with a lowercase letter.");
+      }
+    }
+
+    SQLAdmin sqlAdmin = credentials.getSQLAdmin();
+    String projectId = credentials.getProjectId();
+
+    // Use this list to collect the operations that must reach a RUNNING or DONE state prior to allocate() returning.
+    List<Operation> dbCreationOperations = new ArrayList<Operation>();
+    List<Operation> userCreationOperations = new ArrayList<Operation>();
+    List<String> successfullyCreatedInstancesNames = new ArrayList<String>();
+
+    for (String instanceId : instanceIds) {
+      String decoratedInstanceName = decorateInstanceName(template, instanceId, templateLocalizationContext);
+
+      // Compose the settings.
+      Settings settings = new Settings();
+
+      String tierName = template.getConfigurationValue(TIER, templateLocalizationContext);
+      settings.setTier(tierName);
+
+      IpConfiguration ipConfiguration = new IpConfiguration();
+      ipConfiguration.setIpv4Enabled(true);
+
+      // TODO(kl3n1nz) Might want to whitelist only GCE instances.
```

Comment:
Is this open to the world, or is there some filtering built into the platform that limits visibility?